loadbalancer: Add file-based reflector

joamaki opened the pull request:
As the `cilium-dbg service update` is being removed (https://github.com/cilium/cilium/pull/39084) we need a different mechanism for serving the use-case. 

This implements a file-based synchronization for the load-balancing state by using the Kubernetes Service and EndpointSlice formats, which makes for a simple implementation that is also familiar to users.

joamaki committed:
loadbalancer/reflectors: Implement support for file-based reflection

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer/reflector: Separate conversions to their own file

In preparation for adding another reflector that uses the same conversion
functions, move them to conversions.go and clean up cell.go to refer to
the K8sReflectorCell only.

Add source argument to the conversion functions.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer/reflector: Fix paths to test files in benchmarks

After the code reorg the benchmarks were not updated to point to the
correct files.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/loadbalancer/reflectors/file.go:
```diff
@@ -0,0 +1,230 @@
+)
+
+// StateFile defines the format for the state file used with --lb-state-file.
+type StateFile struct {
```
I'm not sure where I should put this. It is a public API, but at the same time I'm not entirely sure we want to advertise it as such, at least not yet. I'm also not versioning it which might also be a mistake (otoh, implicitly a missing "version" field would imply v1).

joamaki commented on pkg/loadbalancer/reflectors/file.go:
```diff
@@ -0,0 +1,230 @@
+
+		for {
+			select {
+			case ev := <-watcher.Events:
```
This isn't quite right. We should wait for the first event for the file and then wait `LBStateFileInterval` for further events. What happens now is that we might get multiple events about it (especially if editing the file with an editor) and we'd just reload multiple times.

oblazek commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -313,6 +313,7 @@ cilium-agent [flags]
       --l2-pod-announcements-interface-pattern string             Regex matching interfaces used for sending gratuitous arp messages
       --label-prefix-file string                                  Valid label prefixes file path
       --labels strings                                            List of label prefixes used to determine identity of an endpoint
+      --lb-state-file string                                      Synchronize load-balancing state from the specified file
```
do we want this to be a single file thing? Or rather whole folder with multiple *.yaml files? Both options are fine from my side.

joamaki replied:
The benefit with a single file is that then we can have fully atomic changes by swapping the file, but it's of course slightly slower if it has to be updated very quickly and perhaps slightly more inconvenient for some use cases (and this is more convenient for some....).

So I would argue that let's start simple and add support for dividing this into multiple files later if we have a use-case for it. Again I would expect this mechanism to not be used that much once there's e.g. xDS support so I don't think it needs to be made that complex.

oblazek replied:
agree, simple is better

oblazek commented on pkg/loadbalancer/reflectors/conversions.go:
```diff
@@ -0,0 +1,386 @@
+	// The ingress and gateway-api controllers (operator/pkg/model/translation/{gateway-api,ingress}) create
+	// a dummy endpoint to force Cilium to reconcile the service. This is no longer required with this new
+	// control-plane, but due to rolling upgrades we cannot remove it immediately. Hence we have the
+	// special handlind here to just ignore this endpoint to avoid populating the tables with unnecessary
```
nit: typo handlind -> handling

oblazek commented on pkg/loadbalancer/reflectors/conversions.go:
```diff
@@ -0,0 +1,386 @@
+	return headless
+}
+
+func convertService(cfg loadbalancer.ExternalConfig, log *slog.Logger, svc *slim_corev1.Service, source source.Source) (s *loadbalancer.Service, fes []loadbalancer.FrontendParams) {
```
Wondering if a skipped entry should be logged at all or not (as it is now)? Like I am setting up a ipv6 service, but cfg.EnabledIPv6 is not enabled. wdyt?

joamaki replied:
This is used by Cilium in general so starting to suddenly log these is definitely too aggressive. However, we could debug log it. Perhaps makes sense to add debug logs for all the potentially surprising skips.

joamaki replied:
Added debug logs.

oblazek commented on pkg/loadbalancer/reflectors/conversions.go:
```diff
@@ -0,0 +1,386 @@
+
+		state := loadbalancer.BackendStateActive
+		if entry.backend.Terminating {
+			state = loadbalancer.BackendStateTerminating
```
is this on purpose or do we not support other states than active/terminating?

joamaki replied:
From k8s types these are the only 2 states we can reach. The `Quarantined` state can be reached with a health checker that marks a backend as unhealthy (currently there are none in OSS cilium), and the `Maintenance` is currently completely unused (this could've been set via the `cilium-dbg service update` that is now being removed). 

EDIT: Oh wait, we can reach `Maintenance` with a maglev weight of 0. This actually isn't handled in the new control-plane, though it is unclear whether that's needed at all since that can be decided by the reflector. And currently maglev weights can't be set via k8s, so there's no code to change.

oblazek replied:
yeah `Maintenance` is probably usefull only for the standalone mode (to set backend for maintenance), could we support this via annotation potentially?

oblazek replied:
yeah, just wanted to mention weights

joamaki replied:
Yep we could consider adding an annotation for both maintenance and for weights. Were weights needed for your use-case?

oblazek replied:
yeah we have some backends that are not as performant as others

joamaki replied:
Let's add the support for weights in a different PR.

joamaki committed:
loadbalancer/reflector: Debug log dropped frontends/backends

When converting a k8s Service or Endpoints to the internal models
debug log when we ignore data due to configuration or failure to
parse it.

The log.With() is pretty expensive, so we use sync.OnceValue to
lazily do it. Without this we'd be 20% slower in convertService().

This does have a measurable impact on cases where we e.g. have EnableIPv6=false
and lots of services or backends have IPv6 addresses. We'd hit log.With()
and log.Debug() and likely be 20-30% slower. In practice though this wouldn't
really be measurable.

Before:
  BenchmarkConvertService-32               1326844               894.0 ns/op         1118506 services/sec
  BenchmarkConvertEndpoints-32             2550249               452.6 ns/op         2209461 endpoints/sec

After:
  BenchmarkConvertService-32               1241847               905.2 ns/op         1104687 services/sec
  BenchmarkConvertEndpoints-32             3043005               438.2 ns/op         2282010 endpoints/sec

Signed-off-by: Jussi Maki <jussi@isovalent.com>

oblazek reviewed (approved):
LGTM!
